### PR TITLE
Add `zig cc` support for LLVM flag `--no-undefined-version` 

### PR DESCRIPTION
**Pull request by DhashS**
I'm pretty sure this is not all I need to do, so would appreciate some guidance from the broader zig community. 

This issue is what i'm trying to tackle https://github.com/ziglang/zig/issues/16855 by punching that flag through to LLVM.

I've also seen it invoked as `-Wl,--no-undefined-version`, so i'm wondering if to parse it in the .Wl section as well. I didn't initially since it seems like there's a catch-all append call that gets parsed where I specified.

### CODE REVIEW

**Comment by andrewrk** [comment]
Thanks for working on this.

What does this do? The ld --help menu isn't particularly helpful:

```
  --undefined-version         Allow undefined version
  --no-undefined-version      Disallow undefined version
```

File: src/main.zig [hunk 1]
```diff
@@ -2356,6 +2357,8 @@ fn buildOutputType(
                         fatal("unable to parse /version '{s}': {s}", .{ arg, @errorName(err) });
                     };
                     have_version = true;
+                } else if (mem.eql(u8, arg, "--no-undefined-version")) {
+                    linker_no_undefined_version = true;
```
Comment (motiejus) [inline src/main.zig:2361]: Also add handling for `--undefined-version`:

```
                } else if (mem.eql(u8, arg, "--undefined-version")) {
                    linker_no_undefined_version = false;
```

**Comment by motiejus** [comment]
> I'm pretty sure this is not all I need to do, so would appreciate some guidance from the broader zig community.

You will need to add a case in `src/link/Elf.zig`. See a833bdcd7e6fcfee6e9cc33a3f7de78b16a36941 for a simple isolated example.

> I've also seen it invoked as `-Wl,--no-undefined-version`, so i'm wondering if to parse it in the .Wl section as well. I didn't initially since it seems like there's a catch-all append call that gets parsed where I specified.

Looking at the man pages of gcc and clang, I think this is the only valid way to call it (via `-Wl,`). And seems like you did it correctly for that case.

After you've [built zig](https://github.com/ziglang/zig/wiki/Building-Zig-From-Source), you can poke at it:

```
$ ZIG_VERBOSE_LINK=1 zig cc -Wl,--no-undefined-version <...>
```

... and then observe that the flag is present in the line that starts with `ld` (then it's calling the underlying linker).

> Thanks for working on this.
> 
> What does this do? The ld --help menu isn't particularly helpful:
> 
> ```
>   --undefined-version         Allow undefined version
>   --no-undefined-version      Disallow undefined version
> ```

from gcc man page:


           Normally when a symbol has an undefined version, the linker will ignore it. This option disallows symbols with
           undefined version and a fatal error will be issued instead.

Also a relevant lld commit: https://reviews.llvm.org/D135402

**Comment by motiejus** [comment]
Also add a comment in approx https://github.com/ziglang/zig/pull/17326/files#diff-d924ca21d81d7d5a59eb9e10d3e2689c4c58a7e28e6ecce6a064701666f5a730L481

**Comment by DhashS** [comment]
Thanks for the pointer to that commit @motiejus i've done as requested

File: src/Compilation.zig [hunk 2]
```diff
@@ -645,6 +646,8 @@ pub const InitOptions = struct {
     linker_shared_memory: bool = false,
     linker_global_base: ?u64 = null,
     linker_export_symbol_names: []const []const u8 = &.{},
+    linker_undefined_version: ?bool = null,
```
Comment (motiejus) [inline src/Compilation.zig:650]:
```suggestion
```

File: src/Compilation.zig [hunk 3]
```diff
@@ -1586,6 +1589,8 @@ pub fn create(gpa: Allocator, options: InitOptions) !*Compilation {
             .shared_memory = options.linker_shared_memory,
             .global_base = options.linker_global_base,
             .export_symbol_names = options.linker_export_symbol_names,
+            .undefined_version = options.linker_undefined_version,
```
Comment (motiejus) [inline src/Compilation.zig:1593]:
```suggestion
```

File: src/Compilation.zig (hunk 2 above)
Comment (motiejus) [reply src/Compilation.zig:650]: this can be handled in front-end alone without a new field.

File: src/main.zig (hunk 1 above)
Comment (DhashS) [reply src/main.zig:2370]: doing this with a seperate arg to make it a little more explicit

File: src/link/Elf.zig [hunk 4]
```diff
@@ -2054,6 +2054,14 @@ fn linkWithLLD(self: *Elf, comp: *Compilation, prog_node: *std.Progress.Node) !v
             try argv.append(arg);
         }
 
+        if (self.base.options.undefined_version) {
+            try argv.append("--undefined-version");
+        }
+
+        if (self.base.options.no_undefined_version) {
+            try argv.append("--no-undefined-version");
+        }
```
Comment (motiejus) [inline src/link/Elf.zig:2063]:
```suggestion
        if (self.base.options.no_undefined_version) |b| {
          if (b) {
            try argv.append("--no-undefined-version");
          } else {
            try argv.append("--undefined-version");          
          };
        }
```

File: src/link.zig [hunk 5]
```diff
@@ -192,6 +192,8 @@ pub const Options = struct {
     sort_section: ?SortSection,
     major_subsystem_version: ?u32,
     minor_subsystem_version: ?u32,
+    undefined_version: ?bool,
```
Comment (motiejus) [inline src/link.zig:195]:
```suggestion
```

**Review by motiejus — CHANGES_REQUESTED** [review]
I am leaving the suggestion "strange" with double-negatives (i.e. `no_undefined_version` instead of `undefined_version`), because that's the exact flag that `man ld` offers.

File: src/main.zig (hunk 1 above)
Comment (motiejus) [reply src/main.zig:2370]: > doing this with a seperate arg to make it a little more explicit

The surface area in the API should be kept as small as we can make it. The no-`no-` variant should be kept in front-end as per later suggestion.

**Comment by DhashS** [comment]
makes sense, yeah i can't imagine a case where you're specifying both of them and it's correct. 
looks like you left the parsing logic in though, so it'll handle both flags.

**Comment by DhashS** [comment]
Cool, now that i know what you're looking for, i'll give this PR a once over in the morning to make sure it's decent given the back and forth.

File: src/link/Elf.zig [hunk 6]
```diff
@@ -2054,6 +2054,18 @@ fn linkWithLLD(self: *Elf, comp: *Compilation, prog_node: *std.Progress.Node) !v
...
+
+        if (self.base.options.no_undefined_version) |b| {
+          if (b) {
+            try argv.append("--no-undefined-version");
+          } else {
+            try argv.append("--undefined-version");          
+          };
+        }
+
```
Comment (kubkon) [inline src/link/Elf.zig:2068]:
```suggestion
        const allow_undefined_version = self.base.options.allow_undefined_version orelse false;
        if (allow_undefined_version) {
            try argv.append("--undefined-version");
        } else {
            try argv.append("--no-undefined-version");
        }
```

File: src/link.zig [hunk 7]
```diff
@@ -192,6 +192,7 @@ pub const Options = struct {
     sort_section: ?SortSection,
     major_subsystem_version: ?u32,
     minor_subsystem_version: ?u32,
+    no_undefined_version: ?bool,
```
Comment (kubkon) [inline src/link.zig:195]:
```suggestion
    allow_undefined_version: ?bool = null,
```

File: src/main.zig [hunk 8]
```diff
@@ -497,6 +497,8 @@ const usage_build_generic =
     \\  -search_static_only            Only search for static libs.
     \\  -T[script], --script [script]  Use a custom linker script
     \\  --version-script [path]        Provide a version .map file
+    \\  --undefined-version            Allow the linker to link with symbols with undefined version
+    \\  --no-undefined-version         Cause a fatal error if the linker attempts to link to a symbol with an undefined version
```
Comment (kubkon) [inline src/main.zig:501]: Have a look at `lld`'s description which is more accurate as versioning only comes into play from version scripts.

https://github.com/llvm/llvm-project/blob/dc1dc60ea8786d4a2462d157842501d76593f9e7/lld/ELF/Options.td#L480-L482

File: src/main.zig [hunk 9]
```diff
@@ -843,6 +845,8 @@ fn buildOutputType(
     var version_script: ?[]const u8 = null;
     var disable_c_depfile = false;
     var linker_sort_section: ?link.SortSection = null;
+    var linker_undefined_version: ?bool = null;
+    var linker_no_undefined_version: ?bool = null;
```
Comment (kubkon) [inline src/main.zig:849]:
```suggestion
    var linker_allow_undefined_version: ?bool = null;
```

File: src/main.zig [hunk 10]
```diff
@@ -2360,7 +2364,11 @@ fn buildOutputType(
                         fatal("unable to parse /version '{s}': {s}", .{ arg, @errorName(err) });
                     };
                     have_version = true;
-                } else if (mem.eql(u8, arg, "--version")) {
+                } else if (mem.eql(u8, arg, "--undefined-version")) {
+                    linker_undefined_version = true;
```
Comment (kubkon) [inline src/main.zig:2368]:
```suggestion
                    linker_allow_undefined_version = true;
```

File: src/main.zig [hunk 11]
```diff
@@ -2360,7 +2364,11 @@ fn buildOutputType(
...
+                } else if (mem.eql(u8, arg, "--no-undefined-version")) {
+                    linker_no_undefined_version = true;
```
Comment (kubkon) [inline src/main.zig:2370]:
```suggestion
                    linker_allow_undefined_version = false;
```

File: src/Compilation.zig [hunk 12]
```diff
@@ -632,6 +632,8 @@ pub const InitOptions = struct {
     linker_script: ?[]const u8 = null,
     version_script: ?[]const u8 = null,
     soname: ?[]const u8 = null,
+    linker_undefined_version: ?bool = null,
+    linker_no_undefined_version: ?bool = null,
```
Comment (kubkon) [inline src/Compilation.zig:636]:
```suggestion
    linker_allow_undefined_version: ?bool = null,
```

File: src/Compilation.zig [hunk 13]
```diff
@@ -1586,6 +1588,7 @@ pub fn create(gpa: Allocator, options: InitOptions) !*Compilation {
             .shared_memory = options.linker_shared_memory,
             .global_base = options.linker_global_base,
             .export_symbol_names = options.linker_export_symbol_names,
+            .no_undefined_version = options.linker.no_undefined_version,
```
Comment (kubkon) [inline src/Compilation.zig:1591]:
```suggestion
            .allow_undefined_version = options.linker_allow_undefined_version,
```

File: src/main.zig [hunk 14]
```diff
@@ -3428,6 +3436,8 @@ fn buildOutputType(
         .version_script = version_script,
         .disable_c_depfile = disable_c_depfile,
         .soname = resolved_soname,
+        .linker_undefined_version = linker_undefined_version,
+        .linker_no_undefined_version = linker_no_undefined_version,
```
Comment (kubkon) [inline src/main.zig:3440]:
```suggestion
        .linker_allow_undefined_version = linker_allow_undefined_version,
```

**Review by kubkon — CHANGES_REQUESTED** [review]
In addition to my code suggestions, it is also necessary to bump the hash version, and check that everywhere we hit a compile error we have correctly added the flag to relevant caches. Notably [here](https://github.com/ziglang/zig/blob/937e8cb7051a3de537e11c2d52946f772f7449c3/src/Compilation.zig#L2522) and [here](https://github.com/ziglang/zig/blob/937e8cb7051a3de537e11c2d52946f772f7449c3/src/link/Elf.zig#L1996). You should a

**Comment by kubkon** [comment]
I am also thinking that I might start requiring folks enabling new linker features to come up with linker tests and them [here](https://github.com/ziglang/zig/blob/937e8cb7051a3de537e11c2d52946f772f7449c3/test/link/elf.zig#L5-L23). This will help me in the long run as I will be able to simply change the test case from `use_lld = true` to `use_lld = false` and see if my implementation passes the test.

**Comment by DhashS** [comment]
Thanks for the feedback @kubkon, it looks like i misunderstood what @motiejus wanted (i was going to do the opposite, where `no_undefined_version` was the carried-thru variable, but this works fine as well. I have time this weekend to make this submittable from my end, so let me know if there's anything else (i plan to address your comments / links late tomorrow)

**Comment by DhashS** [comment]
In my understanding, things that y'all would like
- [ ] use the LLD docstring
- [ ] make sure caches are happy
- [ ] make build.zig (zig build system) has test coverage for new code paths

**Comment by DhashS** [comment]
okay, not sure how to do those last two.

**Comment by motiejus** [comment]
> In my understanding, things that y'all would like
> 
>     * [x]  use the LLD docstring
> 
>     * [ ]  make sure caches are happy

You are very close. What's missing:

- Add those linker flags to the cache, exactly like Jakub mentioned in https://github.com/ziglang/zig/pull/17326#issuecomment-1741776476 ;
- Bump the constant `link_hash_implementation_version`.

>     * [ ]  make build.zig (zig build system) has test coverage for new code paths

@kubkon I totally see why this would make sense, but I think it's unfair to ask to write such test for newcomers: @DhashS already is shaving the n'th yak by coming from from hermetic_cc_toolchain, because his builds stopped working after Rust upgrade (when it started using this new flag).

Even though I feel like I know *something* in Zig by now (both the language and how it's wired up), I would certainly spend more than a day writing such test. So it would be nice to have, but perhaps too early or much to ask.

**Comment by DhashS** [comment]
sorry, took me a second to figure out what y'all meant.

**Comment by scanterog** [comment]
We're hitting the same issue. Thanks for the fix @DhashS  :) @motiejus @kubkon would you be able to review/merge/release this one 🙏 ?

**Comment by motiejus** [comment]
Looks good to me, but I am not the one deciding what goes to master and when. :)

Anyhow, looks like it is failing CI checks due to compile errors.

File: src/Compilation.zig [hunk 15]
```diff
@@ -2535,6 +2537,8 @@ fn addNonIncrementalStuffToCacheManifest(comp: *Compilation, man: *Cache.Manifes
     man.hash.add(comp.bin_file.options.tsan);
     man.hash.addOptionalBytes(comp.bin_file.options.sysroot);
     man.hash.add(comp.bin_file.options.linker_optimization);
+    man.hash.add(comp.bin_file.options.allow_undefined_version);
```
Comment (scanterog) [inline src/Compilation.zig:2540]:
```suggestion
    man.hash.addOptional(comp.bin_file.options.allow_undefined_version);
```

File: src/link/Elf.zig [hunk 16]
```diff
@@ -2046,6 +2046,7 @@ fn linkWithLLD(self: *Elf, comp: *Compilation, prog_node: *std.Progress.Node) !v
         man.hash.add(self.base.options.tsan);
         man.hash.addOptionalBytes(self.base.options.sysroot);
         man.hash.add(self.base.options.linker_optimization);
+        man.hash.add(self.base.options.allow_undefined_version);
```
Comment (scanterog) [inline src/link/Elf.zig:2049]:
```suggestion
        man.hash.addOptional(self.base.options.allow_undefined_version);
```

**Comment by scanterog** [comment]
yeah I think I've spotted why CI is failing cc @DhashS

**Comment by behos** [comment]
:wave: @kubkon it looks like there are some updates in the PR which might solve the CI issues.
Is there something else blocking this?

**Comment by DhashS** [comment]
would really like to get this merged so that we can bump our rust version :)

**Comment by motiejus** [comment]
All linux tests are failing, it is easier to get attention with a working CI.

**Comment by behos** [comment]
@DhashS looks like a formatting error
```
diff --git a/src/main.zig b/src/main.zig
index a4d7ae93f..9ef095ef6 100644
--- a/src/main.zig
+++ b/src/main.zig
@@ -2379,7 +2379,7 @@ fn buildOutputType(
                     linker_allow_undefined_version = true;
                 } else if (mem.eql(u8, arg, "--no-undefined-version")) {
                     linker_allow_undefined_version = false;
-                }  else if (mem.eql(u8, arg, "--version")) {
+                } else if (mem.eql(u8, arg, "--version")) {
                     try std.io.getStdOut().writeAll("zig ld " ++ build_options.version ++ "\n");
                     process.exit(0);
                 } else {
```

I think this is what's picked up in CI. You can run `zig fmt` on the changed files.

**Pull request closed without merging**